sdl_viewer/src/graphic/moving_window_texture.rs

```diff
@@ -0,0 +1,364 @@
+    /// written with wrapping indexing, so we need to do it ourselves.
+    /// Let's look at the vertical strip:
+    ///
+    /// ```text
+    /// <- width ->
+    /// +---+-+
+    /// | 3 |4|
+    /// +---+-+
+    /// |   | |  vert_strip
+    /// | 1 |2|
+    /// |   | |
+    /// +---+-+
+    /// ```
+    ///
+    /// Wraparound at the right texture edge causes the split into regions 1/3
+    /// vs 2/4, and wraparound at the upper texture edge causes the split into
+    /// regions 1/2 vs 3/4.
+    /// Here's the texture with the vert_strip placed correctly:
+    ///
+    /// ```text
+    /// <- self.size ->
+    /// +-+-------+---+
+    /// | |       |   |
+    /// |2|       | 1 |
+    /// | |       |   |
+    /// +-+       +---+ <- texture_offset.y
+    /// |4|       | 3 |
+    /// +-+-------+---+
+    ///           ^
+    ///           |
+    ///           xoff
+    /// ```
+    ///
+    /// `xoff` is not equal to `texture_offset.x` if the texture moved right.
+    ///
+    /// The way this is handled is by creating regions as if they always need
+    /// splitting, and skipping empty regions that result from that.
+    pub fn incremental_update(
+        &mut self,
+        delta_x: i32,
+        delta_y: i32,
+        vert_strip: ImageBuffer<P, Vec<P::Subpixel>>,
+        hori_strip: ImageBuffer<P, Vec<P::Subpixel>>,
+    ) {
+        // width/height are always positive, but if the delta is negative, the
+        // start of the new region is actually offset + delta, not offset
+        let xoff = if delta_x > 0 {
+            self.u_texture_offset.value.x
+        } else {
+            (self.u_texture_offset.value.x + delta_x).mod_floor(&self.size)
+        };
+        let yoff = if delta_y > 0 {
+            self.u_texture_offset.value.y
+        } else {
+            (self.u_texture_offset.value.y + delta_y).mod_floor(&self.size)
+        };
+
+        // u_texture_offset.value is always in [0; self.size)
+        let x = (self.u_texture_offset.value.x + delta_x).mod_floor(&self.size);
+        let y = (self.u_texture_offset.value.y + delta_y).mod_floor(&self.size);
+        self.u_texture_offset.value = Vector2::new(x, y);
+
+        let vert_regions = UpdateRegion::new_regions(xoff, y, self.size, &vert_strip);
+        let hori_regions = UpdateRegion::new_regions(x, yoff, self.size, &hori_strip);
+        let regions = ArrayVec::from(vert_regions)
+            .into_iter()
+            .chain(ArrayVec::from(hori_regions));
+
+        unsafe {
+            self.gl.BindTexture(opengl::TEXTURE_2D, self.id);
+            for r in regions {
+                let width = i32::try_from(r.pixels.width()).unwrap();
+                let height = i32::try_from(r.pixels.height()).unwrap();
+                if height * width == 0 {
+                    continue;
+                }
+                let image = r.pixels.to_image();
+                self.gl.TexSubImage2D(
+                    opengl::TEXTURE_2D,
+                    0,
```

Comment:
nit: add a comment what this argument is (or use a named constant)? The other arguments are self explanatory.

Reply:
I'd add a link to the documentation, which describes it better than I could. https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glTexImage2D.xhtml

Reply:
I think something like `// level` would be enough in this case + link to a documentation

Reply:
Done